### ISSUE TITLE
Plan: Rust 2024 prelude

### ISSUE THREAD

**Issue by tmandry** [rust-lang/wg-async#310]
**This is a roadmap item for wg-async. You can view the roadmap here: https://github.com/orgs/rust-lang/projects/28**

**Comment by eholk** [comment]
Seems reasonable to me. I'm regularly surprised that `Future` isn't in the prelude.

A lot of code imports `futures::Future`. Would putting `Future` in the prelude break this code?

**Comment by yoshuawuyts** [comment]
> A lot of code imports `futures::Future`. Would putting `Future` in the prelude break this code?

I don't believe so. They both resolve to the same trait anyway, but the way imports work this would Just Work I believe: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=387f434025e1fd5b5625899580464b99)

**Comment by eholk** [comment]
Playing devil's advocate for a bit, are `Future` and `IntoFuture` traits that really need to be in the standard library? Most people don't implement `Future` directly, and probably even fewer call `poll` or `into_future` directly, so maybe it's reasonable to leave these as an advanced feature that people implementing low level async code import but most programmers don't use?

Put another way, what capabilities do we enable by putting them in the prelude?

**Comment by Noah-Kennedy** [comment]
If not in the standard library, then where would they be?

**Comment by compiler-errors** [comment]
> are Future and IntoFuture traits that really need to be in the standard library?

They certainly need to be in the standard library to support `async` and `await` desugaring. 

> Most people don't implement Future directly

Yeah, but I would assume more people do thing like `fn x() -> impl Future<Output = ..>`, `fn x(x: impl Future<Output = ..>)` which would be helped by implementing

I can see how `IntoFuture` probably doesn't need to be in the prelude (since most people are just `.await`ing on things), and I guess I could see `Future` also not being there.

**Comment by eholk** [comment]
Oh sorry, I misspoke! I meant to ask whether we need `Future` and `IntoFuture` in the prelude, not the standard library. I'll edit my comment.

**Comment by eholk** [comment]
> > Most people don't implement Future directly
> 
> Yeah, but I would assume more people using them as impl traits in return or arg position, like `fn x() -> impl Future<Output = ..>`, `fn x(x: impl Future<Output = ..>)` which would be helped by implementing

That's a good point. 

I wonder what the bar is for including things in the prelude is? I'd assume it's for things that essentially all Rust programs will use. Almost every program uses `Option` and iterators, for example. There are still a sizeable number of programs that don't use async though and therefore don't need to use `Future` at all.

**Comment by nrc** [comment]
Yeah, we should talk to the libs team about criteria. Personally, I think we probably don't want to add anything - I reckon a minority of programs involves async code at all and a small minority of those write `impl Future` or implement a future because most just use async and await. Furthermore, those that do use explicit futures tend to just use them in a few places rather than everywhere, so importing them manually doesn't seem onerous. I think this all applies even more strongly to IntoFuture (aiui, the most common usage is implicitly in `await` where it doesn't need to be imported at all)

**Comment by yoshuawuyts** [comment]
> I reckon a minority of programs involves async code at all and a small minority of those write `impl Future` or implement a future because most just use async and await.

This is something we can validate by searching GitHub:

- [`lang:rust /: Future/`](https://github.com/search?q=lang%3Arust+%2F%3A+Future%2F&type=code): yields 40.000 matches
- [`lang:rust /: Iterator/`](https://github.com/search?q=lang%3Arust+%2F%3A+Iterator%2F&type=code): yields 80.000 matches
- [`lang:rust /Future/`](https://github.com/search?q=lang%3Arust+%2FFuture%2F&type=code) yields 700.000 matches
- [`lang:rust /Iterator/`](https://github.com/search?q=lang%3Arust+%2FFuture%2F&type=code) yields 400.000 matches

GitHub code search does have a few limitations. For one the search is not case-sensitive. And it also includes lines which are part of comments. But if we interpret this as a rough indicator rather than a precise measurement, it would seem that `Future` is used about about half as often as iterator in bounds. And likely sees comparable usage elsewhere.

What I'm taking away from this is that the `Future` trait is probably used often enough that it would be justified for inclusion in the stdlib prelude. The most conservative interpretation of this is that `Future` is used about half as much as `Iterator`, one of the most used traits in the stdlib. By-volume that is *a lot*; and it would seem to me as enough to justify its inclusion. `IntoFuture` is most definitely used less than the other traits, but it's also only been part of stable Rust for months, not years like the other traits. At least to me it seems that if we're going to put up `Future` for consideration, `IntoFuture` should also just come with so it would continue mirroring the `Iterator`/`IntoIterator` trait set it was modeled after.

My goal with [this comment](https://github.com/rust-lang/wg-async/issues/310#issuecomment-1488749279) was to specifically request input from the Async WG for which traits (or even types!) we may want to put forward for inclusion. Once we have a rough consensus among ourselves, I was planning to put it forward to the libs team as an ACP to formally clear the acceptance criteria with them.

**Comment by nrc** [comment]
> This is something we can validate by searching GitHub:

I get completely different numbers here and simply clicking these links gives me bogus results (e.g., results with "lang" and "rust" as the found text). I think that at least the repositories searched depends on the user. So I'm not sure these are reliable numbers. Given that GH search also counts forks as different repos, these are also going to be inaccurate that way too. I think a better count might be by downloading the source to the top 500 crates on crates.io and using grep? Or some similar exercise

**Comment by yoshuawuyts** [comment]
> I get completely different numbers here and simply clicking these links gives me bogus results (e.g., results with "lang" and "rust" as the found text)

Oh it may be that you don't yet have access to the [new code search feature](https://github.blog/changelog/2022-11-09-introducing-an-all-new-code-search-and-code-browsing-experience/). It's far more accurate than the previous search, and you can do things like `lang:rust` to scope it just to files written in Rust, and apply regular expressions to look for specific occurrences.

**Comment by samsartor** [comment]
I just stumbled across this post and may actually have the answers available because of another project. Here are the top 50 trait names ranked by average number of explicit mentions in the the top ~1000 Rust projects on GitHub as of November 2022:

<div><div style = "float: left;"></div><div style = "clear: both;"></div></div><div class = "data-frame" style = "overflow-x: scroll;"><table class = "data-frame" style = "margin-bottom: 6px;"><thead><tr class = "header"><th class = "rowNumber" style = "font-weight: bold; text-align: right;"></th><th style = "text-align: left;">Trait Name</th><th style = "text-align: left;">Avg. Mentions per Project</th></tr><tr class = "subheader headerLastRow"></tr></thead><tbody><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">1</td><td style = "text-align: left;">From</td><td style = "text-align: right;">30.6213</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">2</td><td style = "text-align: left;">Display</td><td style = "text-align: right;">18.4828</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">3</td><td style = "text-align: left;">Default</td><td style = "text-align: right;">17.368</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">4</td><td style = "text-align: left;">Debug</td><td style = "text-align: right;">15.9195</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">5</td><td style = "text-align: left;">Send</td><td style = "text-align: right;">5.63669</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">6</td><td style = "text-align: left;">Iterator</td><td style = "text-align: right;">5.58225</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">7</td><td style = "text-align: left;">Clone</td><td style = "text-align: right;">5.37988</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">8</td><td style = "text-align: left;">Drop</td><td style = "text-align: right;">4.52071</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">9</td><td style = "text-align: left;">PartialEq</td><td style = "text-align: right;">4.39408</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">10</td><td style = "text-align: left;">Error</td><td style = "text-align: right;">4.35266</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">11</td><td style = "text-align: left;">AsRef</td><td style = "text-align: right;">3.96686</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">12</td><td style = "text-align: left;">Sync</td><td style = "text-align: right;">3.55503</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">13</td><td style = "text-align: left;">TryFrom</td><td style = "text-align: right;">3.52899</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">14</td><td style = "text-align: left;">Deref</td><td style = "text-align: right;">2.91006</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">15</td><td style = "text-align: left;">Fn</td><td style = "text-align: right;">2.65917</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">16</td><td style = "text-align: left;">IntoIterator</td><td style = "text-align: right;">2.59408</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">17</td><td style = "text-align: left;">Into</td><td style = "text-align: right;">2.33018</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">18</td><td style = "text-align: left;">FromStr</td><td style = "text-align: right;">2.1787</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">19</td><td style = "text-align: left;">Eq</td><td style = "text-align: right;">2.15621</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">20</td><td style = "text-align: left;">Serialize</td><td style = "text-align: right;">2.11243</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">21</td><td style = "text-align: left;">Future</td><td style = "text-align: right;">1.7645</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">22</td><td style = "text-align: left;">Hash</td><td style = "text-align: right;">1.72071</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">23</td><td style = "text-align: left;">PartialOrd</td><td style = "text-align: right;">1.58225</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">24</td><td style = "text-align: left;">FnMut</td><td style = "text-align: right;">1.57988</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">25</td><td style = "text-align: left;">Write</td><td style = "text-align: right;">1.5432</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">26</td><td style = "text-align: left;">Deserialize</td><td style = "text-align: right;">1.52899</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">27</td><td style = "text-align: left;">Copy</td><td style = "text-align: right;">1.40355</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">28</td><td style = "text-align: left;">FnOnce</td><td style = "text-align: right;">1.3716</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">29</td><td style = "text-align: left;">Ord</td><td style = "text-align: right;">1.20592</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">30</td><td style = "text-align: left;">DerefMut</td><td style = "text-align: right;">1.17633</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">31</td><td style = "text-align: left;">Stream</td><td style = "text-align: right;">1.12071</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">32</td><td style = "text-align: left;">Parse</td><td style = "text-align: right;">0.942012</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">33</td><td style = "text-align: left;">Add</td><td style = "text-align: right;">0.926627</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">34</td><td style = "text-align: left;">Unpin</td><td style = "text-align: right;">0.822485</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">35</td><td style = "text-align: left;">Index</td><td style = "text-align: right;">0.790533</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">36</td><td style = "text-align: left;">Read</td><td style = "text-align: right;">0.76568</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">37</td><td style = "text-align: left;">ExactSizeIterator</td><td style = "text-align: right;">0.757396</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">38</td><td style = "text-align: left;">Visitor</td><td style = "text-align: right;">0.753846</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">39</td><td style = "text-align: left;">Encode</td><td style = "text-align: right;">0.659172</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">40</td><td style = "text-align: left;">Sealed</td><td style = "text-align: right;">0.63787</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">41</td><td style = "text-align: left;">Sub</td><td style = "text-align: right;">0.63432</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">42</td><td style = "text-align: left;">Deserializer</td><td style = "text-align: right;">0.598817</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">43</td><td style = "text-align: left;">Sized</td><td style = "text-align: right;">0.589349</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">44</td><td style = "text-align: left;">AsyncRead</td><td style = "text-align: right;">0.583432</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">45</td><td style = "text-align: left;">Mul</td><td style = "text-align: right;">0.571598</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">46</td><td style = "text-align: left;">FromIterator</td><td style = "text-align: right;">0.569231</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">47</td><td style = "text-align: left;">ToTokens</td><td style = "text-align: right;">0.564497</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">48</td><td style = "text-align: left;">AsyncWrite</td><td style = "text-align: right;">0.551479</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">49</td><td style = "text-align: left;">Arbitrary</td><td style = "text-align: right;">0.531361</td></tr><tr><td class = "rowNumber" style = "font-weight: bold; text-align: right;">50</td><td style = "text-align: left;">Service</td><td style = "text-align: right;">0.478107</td></tr></tbody></table></div>

Based on this,  `Display` and `Debug` should really be in the prelude.`Error` is also pretty far up the list. But `Future` is on the cusp, just behind serde's `Serialize` trait (obviously not in the prelude) but ahead of some prelude traits like `Ord`, `PartialOrd`, `FnMut`, and `Copy`.

We can also do an over-time comparison between `Future` and other similar traits, such as `Iterator`. For a while it looked as if people would soon be mentioning the `Future` trait almost as much as  `Iterator`, but the gap has widened since async/await was stabilized: 
![image](https://user-images.githubusercontent.com/1740180/233484443-8913eeba-360a-43f9-9266-b149cf1324d4.png)

But looking at only projects with at least one usage of either, we can see that projects which write out `Future` at all tended to write it out a lot, almost more so than a project which similarly mentions the Iterator trait:
![image](https://user-images.githubusercontent.com/1740180/233483269-0dd92dc7-1402-4e2e-a027-4ba5bbc6d0aa.png)

This obviously doesn't compare usage of the respective `Iterator::*` and `Future::*` methods when the traits themselves are unnamed (doing trait resolution was well beyond the project scope), but these plots do support the idea that `pub use Future` should be in the prelude as much as `pub use Iterator` if not as much as `pub use Iterator as _`.

We can also break down these mentions of the `Future` trait by their position in the syntax, to see that people are writing manual `impl Future for T` impls surprisingly often. The `fn() -> impl/dyn Future` was on the rise until async/await, then was largely replaced. On the other hand, use of `f(_: impl/dyn Future)` has been steadily increasing:
![image](https://user-images.githubusercontent.com/1740180/233563215-2a511fad-cd7d-478a-8bc1-72a7c9271034.png)

Compare this to the `Iterator` trait, where `impl Iterator for T` mentions dominate to a greater extent, and where the relative growth of the different syntaxes has remained fairly proportional after the `-> impl Iterator` syntax was stabilized:
![image](https://user-images.githubusercontent.com/1740180/233565028-b4ed9ff0-a22f-4a93-a1e6-3bd326f3c32b.png)

We can also investigate why people still type out `fn() -> impl Future` at all, when async is available. It is about half as popular as before the async syntax was stabilized. But notice that people customize the lifetimes on returned futures just as much as before:
![image](https://user-images.githubusercontent.com/1740180/233485303-fcee58aa-c008-41a7-9d4b-e18500e15e7d.png)

My take from this is that having Future in the prelude would be nice for a lot of projects even if async/await has reduced that need somewhat. However, other traits might be more important to add.

**Comment by taiki-e** [comment]
- Since `task::{Context, Poll}` is also needed to implement `Future`, including `Future` in the prelude would be of little use from the standpoint of eliminating import effort during manual implementation. (This also applies to `Display` and `Debug` mentioned by @samsartor, which need other items in the fmt module for implementation.)

- [The data](https://github.com/rust-lang/wg-async/issues/310#issuecomment-1491712570) provided by @yoshuawuyts does not appear to be very reliable, as we can see cases from the first page that clearly match irrelevant items (e.g., "S: futures::Stream" is considered to contain one ": Future", or "use futures::future::FutureExt;" is considered to contain three "Future"'s.). It should also be noted that the GitHhub search also searches forks by default.

- If other async-related traits are added to the std (e.g., `Stream`/`AsyncIterator`), do we want to include them in the prelude as well?
  - If yes, should we repeat the same discussion in 3 years, or include it in prelude at stabilization?
  - If no, why is `Future` trait special and the others are not?

- Other traits that are not included in `std::prelude` but are widely used are `std::io` traits, which have module-level prelude (`std::io::prelude`). The "only io-related code needs it" nature of io traits is somewhat similar to async-related traits, where "only async-related code needs it". Maybe we may want to consider the idea regarding module-level prelude (something like `std::future::prelude::{Future, AsyncIterator}`).

**Comment by yoshuawuyts** [comment]
The [2024 prelude](https://doc.rust-lang.org/nightly/std/prelude/rust_2024/index.html) is now available behind a flag on nightly. We should consider PR'ing `Future` and `IntoFuture` to it.
